### ISSUE TITLE
JSON praser can't take an int128

### ISSUE THREAD

**Issue by meatball133** [crystal-lang/crystal#12953]
## Bug Report

When giving json parser a json with int127 so does the crystal json.praser return an error:

```crystal
Unhandled exception: Invalid Int64: "186709961001538790100634132976990" at line 94, column 19 (JSON::ParseException)
```

**Comment by meatball133** [comment]
Changing this to i128, would that fix it? https://github.com/crystal-lang/crystal/blob/c6e3116df9a44a871740318ebcf486f986b79dad/src/json/pull_parser.cr#L212

**Comment by straight-shoota** [comment]
Not quire, but it would be one step in that direction. However, it would also be a breaking change.

We also need to represent the type in `JSON:Any`. And then there's also the problem that `Int128` isn't really the end. JSON number can be arbitrarily large.

Maybe something like https://github.com/crystal-lang/crystal/issues/11490#issuecomment-1165610021 would be more appropriate. Or require `BigInt` for this.

**Comment by meatball133** [comment]
For my use case I wont need bigger than Int128. But yeah that could be a limitation, I am open for all suggestions on how to solve this. One solution could be to let the user (programmer decide), but otherwise, a BigInt is also an option.

**Comment by straight-shoota** [comment]
Support for large numbers is also incomplete for `JSON::Serializable`. I think that would be the first and easiest step. It requires some improvements to the parser, which should be easy to implement in a backwards-compatible way. This would already help work with large numbers when the data format is known.

For `JSON::Any`, it's probably best to represent large numbers as `String`. `Int128` may suffice for your use case, but numbers can be bigger in JSON. This should be backwards-compatible. We'll probably need add an ivar to indicate the original JSON type, though. But that should be fine.
I think this solution should be doable without breaking anything and it's powerful and flexible.

**Comment by meatball133** [comment]
You mean like if I have this number (we say it is bigger than int64::max) 100, we should parse it as "100_int128" if it is possible to fit it as a int128. And that we have something like "100_int_big" for numbers bigger than that. Or do you mean that all numbers bigger than int64 should be parsed as like "100_int"? Or do you mean in a totally different way?

**Comment by straight-shoota** [comment]
No, I mean when the parser encounters a number that can't be represented as Int64 or Float64, it parses it *literally* as a string and attaches the original JSON type.

`JSON.parse("186709961001538790100634132976990")` would return a `JSON::Any` with raw value `"186709961001538790100634132976990"` and an additional type field that identifies it as `integer`.
Then you can parse that value as Int128 or BigDecimal or whatever type you want at call site.

The need for an additional type identifier may be debatable. Its only purpose is to ensure type safety i.e. to make a difference between a JSON string (`"186709961001538790100634132976990"`) and number (`186709961001538790100634132976990`). This is often probably not really necessary.

**Comment by HertzDevil** [comment]
A JSON string this long could be a hexstring that happens to contain decimal digits only. In this case it would be odd if this is readable as an integer, but fails as soon as there is an `/[A-Fa-f]/` somewhere

**Comment by straight-shoota** [comment]
Sure, it could be anything. It's just a string that contains only digits. It doesn't need to be interpreted as a number.
I think it's a viable option to do this interpretation at the call site. Whether you call `as_s` or `as_i128` (or whatever).

Additional type safety surely wouldn't be bad, but I'm not sure how relevant it is. The line is already quite blurred because of the fact that many JSON parsers actually can't deal with large numbers and thus it's relatively common to serialize large numbers as strings.

**Comment by asterite** [comment]
If I parse `{"a": "123"}` into JSON and then get the node for "123" and call `as_i`, I'd really want that to fail. It's not an int, it's a string.

So I think it would be best to have an extra ivar that tells whether the string is actually a number or not.

Not doing that would also be a breaking change, because code that used to fail could now silently start working fine.

**Comment by meatball133** [comment]
`:int_as_string` as ivar?

**Comment by straight-shoota** [comment]
It's probably best to make the ivar hold a generic `json_type`. We can probably reuse `JSON::Token::Kind` for this.

**Comment by straight-shoota** [comment]
> If I parse `{"a": "123"}` into JSON and then get the node for "123" and call `as_i`, I'd really want that to fail. It's not an int, it's a string.

Yeah, that should never succeed. But that's not really the subject here. We're talking about big numbers that don't fit in `Int64`, so we'd need new methods for additional types such as Int128 and BigInt. And we could introduce them to work on ordinary string values without breaking anything.

**Comment by meatball133** [comment]
Well I made an attempt, I am unsure if it is how you guys imagined it, but I did some testing and it doesn't work for numbers that is int64 (or lower) or strings.

**Comment by HertzDevil** [comment]
If we are only concerned with `JSON::Serializable`, we could easily circumvent `JSON::PullParser`'s typed interface by relying on `#read_raw` inside a converter directly:

```crystal
require "json"
require "big"

module IntConverter(T)
  def self.from_json(pull : JSON::PullParser) : T
    {% raise "Expected T to be a non-union integer type, got #{T}" unless T < Int && T.union_types.size == 1 %}
    pull.raise "Expected #{JSON::PullParser::Kind::Int} but was #{pull.kind}" unless pull.kind.int? # pull.expect_kind :int
    T.new(pull.read_raw)
  end

  def self.to_json(value : T, json : JSON::Builder) : Nil
    {% raise "Expected T to be a non-union integer type, got #{T}" unless T < Int && T.union_types.size == 1 %}
    json.raw(value.to_s)
  end
end

class Foo
  include JSON::Serializable

  @[JSON::Field(converter: IntConverter(Int128))]
  getter x : Int128

  @[JSON::Field(converter: IntConverter(UInt64))]
  getter y : UInt64

  @[JSON::Field(converter: IntConverter(BigInt))]
  getter z : BigInt
end

f = Foo.from_json <<-JSON
{
  "x": #{Int128::MIN},
  "y": #{UInt64::MAX},
  "z": #{10.to_big_i ** 50}
}
JSON
f         # => Foo(@x=-170141183460469231731687303715884105728, @y=18446744073709551615, @z=100000000000000000000000000000000000000000000000000)
f.to_json # => %({"x":-170141183460469231731687303715884105728,"y":18446744073709551615,"z":100000000000000000000000000000000000000000000000000})
```

Then the `PullParser` interface improvements become relevant only to `JSON::Any`. Ditto for floats and YAML.
